components/butterfly/src/network.rs

```diff
@@ -0,0 +1,297 @@
+// Network trait is used (trace, expire, inbound, outbound and so
+// on). I added it here, because Network is exclusively used by the
+// butterfly component.
+//
+// Same for Debug - Network is used by the butterfly component only so
+// I add it here to save me some typing.
+/// A trait for types used to provide swim and gossip sockets.
+pub trait Network: Send + Sync + Debug + 'static {
+    type SwimSocket: SwimSocket;
+    type GossipPushSocket: GossipPushSocket;
+    type GossipPullSocket: GossipPullSocket;
+
+    fn get_swim_addr(&self) -> SocketAddr;
+    fn get_swim_port(&self) -> u16;
+    fn get_swim_socket(&self) -> Result<Self::SwimSocket>;
+
+    fn get_gossip_addr(&self) -> SocketAddr;
+    fn get_gossip_port(&self) -> u16;
+    fn get_gossip_push_socket(&self, addr: SocketAddr) -> Result<Self::GossipPushSocket>;
+    fn get_gossip_pull_socket(&self) -> Result<Self::GossipPullSocket>;
+}
+
+/// An implementation of SwimSocket that uses UdpSocket.
+#[derive(Debug)]
+pub struct UdpSwimSocket {
+    udp: UdpSocket,
+}
+
+impl TryClone for UdpSwimSocket {
+    fn try_clone(&self) -> Result<Self> {
+        Ok(UdpSwimSocket {
+            udp: self.udp.try_clone().map_err(|_| Error::SocketCloneError)?,
+        })
+    }
+}
+
+impl SwimSocket for UdpSwimSocket {
+    fn send_to(&self, buf: &[u8], addr: SocketAddr) -> Result<usize> {
+        self.udp.send_to(buf, addr).map_err(
+            |e| Error::SwimSendError(e),
+        )
+    }
+
+    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, SocketAddr)> {
+        self.udp.recv_from(buf).map_err(
+            |e| Error::SwimReceiveError(e),
+        )
+    }
+}
+
+/// An implementation of GossipPushSocket and GossipPullSocket that
+/// uses zmq::Socket.
+pub struct ZmqGossipSocket {
+    zmq: zmq::Socket,
+}
+
+impl GossipPushSocket for ZmqGossipSocket {
+    fn send(&self, buf: &[u8]) -> Result<()> {
+        self.zmq.send(buf, 0).map_err(|e| {
+            Error::GossipPushSocketSendError(e.description().to_owned())
+        })
+    }
+}
+
+impl GossipPullSocket for ZmqGossipSocket {
+    fn recv(&self) -> Result<Vec<u8>> {
+        self.zmq.recv_bytes(0).map_err(|e| {
+            Error::GossipPullSocketReceiveError(e.description().to_owned())
+        })
+    }
+}
+
+/// An implementation of Network that creates UdpSwimSocket for SWIM
+/// communication, and ZmqGossipSockets for gossip communication.
+pub struct RealNetwork {
+    swim_addr: SocketAddr,
+    gossip_addr: SocketAddr,
+    push_socket_linger: i32,
+    zmq_context: ServerContext,
+}
+
+impl RealNetwork {
+    /// Create an instance of RealNetwork to be used by clients. It
+    /// sets linger time for push zmq gossip sockets to
+    /// indefinite. Use this instance to get pull or push gossip
+    /// sockets.
+    ///
+    /// For getting swim sockets, get an instance with
+    /// new_for_server().
+    pub fn new_for_client() -> Self {
+        Self::new(Self::throw_away_addr(), Self::throw_away_addr(), -1)
```

Comment:
Why does it use "throw-away" addresses? Maybe a comment could explain that.

Reply:
Client only sends gossip messages (rumors) and the target of the message is specified in the get_gossip_push_socket function. The addresses here are for receiving/sending SWIM messages or receiving rumors, which client does not do.

Will add an explanation there.